core/src/main/java/org/radargun/utils/Fuzzy.java

```diff
@@ -51,57 +55,69 @@ public String toString() {
-         if (weight <= 0) throw new IllegalArgumentException();
+      public Builder addWeighted(T value, BigDecimal weight) {
+         //less then 0
+         if (weight.compareTo(BigDecimal.ZERO) <= 0) throw new IllegalArgumentException();
          weightedValues.add(value);
          weights.add(weight);
          return this;
       }
 
-      public Builder addFixed(T value, double probability) {
-         if (probability <= 0 || probability > 1) throw new IllegalArgumentException();
+      public Builder addFixed(T value, BigDecimal probability) {
+         if (!isValid(probability)) throw new IllegalArgumentException();
          fixedValues.add(value);
          probabilities.add(probability);
          return this;
       }
 
+      private boolean isValid(BigDecimal probability) {
+         //smaller than 0 or greater than 1
+         if (probability.compareTo(BigDecimal.ZERO) <= 0 || probability.compareTo(BigDecimal.ONE) == 1)
+            return false;
+         else
+            return true;
+      }
+
       public Fuzzy<T> create() {
          if (weightedValues.size() + fixedValues.size() == 0) throw new IllegalStateException();
-         double sumWeight = 0;
-         for (Double w : this.weights) {
-            sumWeight += w;
+         BigDecimal sumWeight = BigDecimal.ZERO;
+         for (BigDecimal w : this.weights) {
+            sumWeight = sumWeight.add(w);
          }
-         double[] cumulativeProbability = new double[this.probabilities.size() + this.weights.size()];
+         BigDecimal[] cumulativeProbability = new BigDecimal[this.probabilities.size() + this.weights.size()];
          int i = 0;
-         double cumulatedProbability = 0;
-         for (Double p : this.probabilities) {
-            cumulatedProbability += p;
+         BigDecimal cumulatedProbability = BigDecimal.ZERO;
+         for (BigDecimal p : this.probabilities) {
+            cumulatedProbability = cumulatedProbability.add(p);
             cumulativeProbability[i++] = cumulatedProbability;
          }
-         if (cumulatedProbability > 1
-            || (cumulatedProbability == 1 && weights.size() > 0)
-            || (cumulatedProbability < 0.99999 && weights.size() == 0)) {
+         if (cumulatedProbability.compareTo(BigDecimal.ONE) == 1 //bigger than 1
+            || (cumulatedProbability.compareTo(BigDecimal.ONE) == 0 && weights.size() > 0) //equal to one
+            || (cumulatedProbability.compareTo(BigDecimal.ONE) == -1 && weights.size() == 0)) { //less than one
             throw new IllegalStateException("Probability: " + cumulatedProbability);
          }
-         double cumulatedWeight = 0;
-         for (Double w : this.weights) {
-            cumulatedWeight += w;
-            cumulativeProbability[i++] = (1 - cumulatedProbability) * cumulatedWeight / sumWeight;
+         BigDecimal cumulatedWeight = BigDecimal.ZERO;
+         MathContext mc = new MathContext(2, RoundingMode.HALF_UP); //required for division
```

Comment:
Rounding is required because some results of division can have indefinite number of decimal points. Anyway, I still set the last cumulativeProbability to 1.0 manually because I had this concern too.